### PR TITLE
Postgres storage

### PR DESCRIPTION
**Pull request by antmat**
@3hren PTAL, but this is still WIP. Things to do:
- More logs
- Code style
- Create special pqxx category and convert pqxx exceptions to system_error
- Support for plain PgSQL storage without underlying storage
- Move column names to config

### CODE REVIEW

**Comment by antmat** [comment]
@shaitan Maybe, you are interested too

File: postgres/src/pool.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            auto slot = std::make_shared<slot_t>(io_loop, connection_string);
+            auto id = slot->id();
+            _slots.emplace(id, std::move(slot));
+        }
+    });
+}
+
+pool_t::~pool_t() {
+    io_work.reset();
+    slots->clear();
```
Comment (3Hren) [inline postgres/src/pool.cpp:23]: Not necessary.

File: postgres/src/postgres.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+namespace cocaine {
+namespace storage {
+
+postgres_t::postgres_t(context_t& context, const std::string& name, const dynamic_t& args) :
+    api::storage_t(context, name, args),
+    table_name(args.as_object().at("pg_table_name", "cocaine_index").as_string()),
+    wrapped(api::storage(context, args.as_object().at("pg_underlying_storage", "core").as_string())),
+    pg_pool(args.as_object().at("pg_pool_size", 1ul).as_uint(),
+            args.as_object().at("pg_connection_string", "").as_string())
```
Comment (3Hren) [inline postgres/src/postgres.cpp:24]: Without connection string it's useless, even with the empty one.

Maybe better to explicitly check the `pg_connection_string` sanity in the ctor body?

File: postgres/src/postgres.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            cb(make_exceptional_future<void>());
+        }
+    });
+}
+
+void
+postgres_t::remove(const std::string& collection, const std::string& key, callback<void> cb) {
+    wrapped->remove(collection, key, [=](std::future<void> fut){
+        try {
+            // First we check if underlying write succeed
```
Comment (3Hren) [inline postgres/src/postgres.cpp]: s/write/remove

**Review by 3Hren — CHANGES_REQUESTED** [review]
Looks almost good for me. Add newlines at the end of files, fix minor issues and we can merge.

**Comment by 3Hren** [comment]
Also fix debian/control.

File: postgres/src/postgres.cpp (hunk 2 above)
Comment (antmat) [reply postgres/src/postgres.cpp:24]: empty string imply default connection param via local socket

Comment (3Hren) [reply postgres/src/postgres.cpp:24]: Ok then.

File: postgres/src/pool.cpp (hunk 1 above)
Comment (antmat) [reply postgres/src/pool.cpp:23]: I prefer explicit destruction of slots here. There is more control

File: debian/libcocaine-plugin-postgres-dev.install [hunk 4]
```diff
@@ -0,0 +1 @@
+usr/include/cocaine/error/postgres.hpp
```
Comment (3Hren) [inline debian/libcocaine-plugin-postgres-dev.install]: Da ti zaebal!

**Review by 3Hren — CHANGES_REQUESTED** [review]

File: debian/libcocaine-plugin-postgres-dev.install (hunk 4 above)
Comment (antmat) [reply debian/libcocaine-plugin-postgres-dev.install]: wat?

**Review by 3Hren — APPROVED** [review]

**Pull request merged**